src/main/QueryServer.h

```diff
@@ -0,0 +1,50 @@
+#pragma once
+
+// Copyright 2024 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "lib/httpthreaded/server.hpp"
+
+#include <functional>
+#include <map>
+#include <memory>
+#include <string>
+#include <thread>
+
+namespace stellar
+{
+class SearchableBucketListSnapshot;
+class BucketSnapshotManager;
+
+class QueryServer
+{
+  private:
+    typedef std::function<bool(QueryServer*, std::string const&,
+                               std::string const&, std::string&)>
+        HandlerRoute;
+
+    httpThreaded::server::server mServer;
+
+    std::map<std::thread::id, std::shared_ptr<SearchableBucketListSnapshot>>
+        mBucketListSnapshots;
+
+    bool safeRouter(HandlerRoute route, std::string const& params,
+                    std::string const& body, std::string& retStr);
+
+    bool notFound(std::string const& params, std::string const& body,
+                  std::string& retStr);
+
+    void addRoute(std::string const& name, HandlerRoute route);
+
+    // Returns raw LedgerKeys for the given keys from the Live BucketList. Does
+    // not query other BucketLists or reason about archival.
+    bool getLedgerEntryRaw(std::string const& params, std::string const& body,
+                           std::string& retStr);
+
+  public:
+    QueryServer(const std::string& address, unsigned short port, int maxClient,
+                size_t threadPoolSize,
```

Comment:
I think we'll have to do more measurements here later, but I think we might be ok because we have distinct work classes for each thread pool. Overlay should be mostly networking bound. Bucket merges, proof production, and this query server is mostly disk bound, and parallel apply is all CPU because we preflight everything before hitting the VM. I think we should be able to spin up a large number of overlay and BucketDB threads without interfering too much with apply, but more testing will be required.